include/index_helper.cuh

```diff
@@ -1106,4 +1106,302 @@ namespace quda {
+     *
+     * @return     cbs
+     */
+    __device__ __host__ inline int setup_cbs(const int mu, const int *X)
+    {
+      if (mu==0) {
+        return X[0];
+      } else if ((X[mu]%4)==0) {
+        return 4;
+      } else if ((X[mu]%3)==0) {
+        return 3;
+      } else if ((X[mu]%2)==0) {
+        return 2;
+      } else {
+        return 1;
+      }
+    }
+
+
+    /**
+     * @brief      Pure function to return ipt[iy], where
+     *             iy=x3+L3*x2+L2*L3*x1+L1*L2*L3*x0 without accessing the
+     *             ipt-array, but calculating the index on the fly. Notice that xi
+     *             and Li are in openQCD (txyz) convention. If they come from
+     *             QUDA, you have to rotate them first (see rotate_coords).
+     *
+     * @param[in]  x     Carthesian local lattice corrdinates, 0 <= x[i] < L[i]
+     * @param[in]  L     Local lattice extents
+     *
+     * @return     ipt[x3+L3*x2+L2*L3*x1+L1*L2*L3*x0] = the local flat index of
+     *             openQCD
+     */
+    __device__ __host__ inline int ipt(const int *x, const int *L)
+    {
+      int xb[4], xn[4];
+
+      // openQCDs cache block size in txyz convention
+      int cbs[4] = {setup_cbs(0, L), setup_cbs(1, L), setup_cbs(2, L), setup_cbs(3, L)};
+
+      // openQCDs cache block grid in txyz convention
+      int cbn[4] = {L[0]/cbs[0], L[1]/cbs[1], L[2]/cbs[2], L[3]/cbs[3]};
+
+      xb[0] = x[0] % cbs[0];
+      xb[1] = x[1] % cbs[1];
+      xb[2] = x[2] % cbs[2];
+      xb[3] = x[3] % cbs[3];
+
+      xn[0] = x[0]/cbs[0];
+      xn[1] = x[1]/cbs[1];
+      xn[2] = x[2]/cbs[2];
+      xn[3] = x[3]/cbs[3];
+
+      return (
+         lexi(xb, cbs, 4)/2
+       + vol(cbs)*lexi(xn, cbn, 4)/2
+       + ((x[0]+x[1]+x[2]+x[3]) % 2 != 0)*(vol(L)/2) // odd -> +VOLUME/2
+      );
+    }
+
+
+    /**
+     * @brief      Determines the number of boundary points in direction mu prior to
+     *             the Carthesian index x with dimensions X
+     *
+     * @param[in]  mu    Direction
+     * @param[in]  x     Lattice point
+     * @param[in]  X     Dimensions of the lattice/block
+     *
+     * @return     Number of prior boundary points
+     */
+    __device__ __host__ inline int boundary_pts(const int mu, const int *x, const int *X)
+    {
+      int ret = 0;
+
+      if (mu==3) {
+        ret = lexi(x, X, 3); // lexi without x[3]
+      } else if (mu==2) {
+        ret = X[3]*lexi(x, X, 2);
+        if (x[2]==(X[2]-1)) {
+          ret += x[3]; // lexi without x[2]
+        }
+      } else if (mu==1) {
+        if (x[1]==(X[1]-1)) {
+          ret = X[2]*X[3]*x[0] + X[3]*x[2] + x[3]; // lexi without x[1]
+        } else {
+          ret = surface(X, 1);
+        }
+      } else if (mu==0) {
+        if (x[0]==(X[0]-1)) {
+          ret = lexi(x+1, X+1, 3); // lexi without x[0]
+        } else {
+          ret = surface(X, 0);
+        }
+      }
+
+      return ret;
+    }
+
+
+    /**
+     * @brief      Pure implementation of iup[ix][mu]
```

Comment:
Maybe add: Returns neighbouring point in positive $\mu$ direction.

Reply:
Comment added